src/cpp/tensorbase.cpp

```diff
@@ -42,130 +43,151 @@ TensorBase::TensorBase(const std::string& name,
+        delete(reinterpret_cast<unsigned char *>(_data));
+        _data = NULL;
+    }
 }
 
+// TensorBase copy assignment operator
 TensorBase& TensorBase::operator=(const TensorBase& tb)
 {
-    /* This is the copy assignment operator for
-    TensorBase.  A deep copy of the tensor
-    data is performed.
-    */
-   if(this!=&tb) {
-        this->_name = tb._name;
-        this->_type = tb._type;
-        this->_dims = tb._dims;
-        if(this->_data)
-            free(this->_data);
-   }
+    // check for self-assignment
+    if (&tb == this)
+        return *this;
+
+    // deep copy tensor data
+    _name = tb._name;
+    _type = tb._type;
+    _dims = tb._dims;
+
+    // Erase our old data
+    if (_data != NULL) {
+        delete(reinterpret_cast<unsigned char *>(_data));
+        _data = NULL;
+    }
+
+    // NOTE: The actual tensor data will be copied by the child class
+    // (template) after it calls this assignment operator.
+    // This routine should never be called directly
+
+    // Done
    return *this;
 }
 
+// TensorBase move assignment operator
 TensorBase& TensorBase::operator=(TensorBase&& tb)
 {
-    /* This is the move assignment operator for
-    TensorBase.
-    */
-    if(this!=&tb) {
-        this->_name = std::move(tb._name);
-        this->_type = std::move(tb._type);
-        this->_dims = std::move(tb._dims);
-        if(this->_data)
-            free(this->_data);
-        this->_data = tb._data;
-        tb._data = 0;
-    }
+    // check for self-assignment
+    if (&tb == this)
+        return *this;
+
+    // Move tensor data
+    _name = std::move(tb._name);
+    _type = std::move(tb._type);
+    _dims = std::move(tb._dims);
+
+    // Erase our old data and assume ownership of tb's data
+    if (_data != NULL)
+        delete(reinterpret_cast<unsigned char *>(_data));
+    _data = tb._data;
+    tb._data = NULL;
+
+    // Done
     return *this;
 }
 
+// Retrieve the tensor name.
 std::string TensorBase::name()
 {
-    /* Return the tensor name.
-    */
-    return this->_name;
+    return _name;
 }
 
+// Retrieve the tensor type.
 TensorType TensorBase::type()
 {
-    /* Return the tensor type.
-    */
-   return this->_type;
+   return _type;
 }
 
-std::string TensorBase::type_str(){
-    /* Return the string version
-    of the tensor type.
-    */
-    return TENSOR_STR_MAP.at(this->type());
+// Retrieve the string version of the tensor type.
+std::string TensorBase::type_str()
+{
+    return TENSOR_STR_MAP.at(type());
 }
 
+// Retrieve the tensor dims.
 std::vector<size_t> TensorBase::dims()
 {
-    /* Return the tensor dims
-    */
-   return this->_dims;
+   return _dims;
 }
 
+// Retrieve the total number of values in the tensor.
 size_t TensorBase::num_values()
 {
-    /* Return the total number of values in the tensor
-    */
-    size_t n_values = this->_dims[0];
-    for(size_t i=1; i<this->_dims.size(); i++) {
-        n_values *= this->_dims[i];
-    }
+    size_t n_values = 1;
```

Comment:
On one hand I can see how this is better, but I think the previous is slightly better because an error would be thrown if ``_dims.size()`` is 0 (bad state), whereas the second would just return ``1`` which is not true and misleading.